core/packages/contracts/contracts/NativeTokens.sol

```diff
@@ -0,0 +1,166 @@
+contract NativeTokens is Ownable {
+    /// @dev Describes the type of message.
+    enum Action {
+        /// @dev The default value for messages created.
+        Uninitialized,
+        /// @dev A message which unlocks funds for native tokens.
+        Unlock
+    }
+
+    /// @dev Message format.
+    struct Message {
+        /// @dev The action type.
+        Action action;
+        /// @dev The message payload.
+        bytes payload;
+    }
+
+    /// @dev Unlock payload format.
+    struct UnlockPayload {
+        /// @dev The ERC20 token to unlock.
+        address token;
+        /// @dev The destination address that will receive unlocked funds.
+        address recipient;
+        /// @dev The amount to unlock.
+        uint128 amount;
+    }
+
+    /// @dev Emitted once the funds are locked and a message is successfully queued.
+    /// @param origin The address which initiated the lock.
+    /// @param recipient The substrate address that will receive the funds.
+    /// @param token The token locked.
+    /// @param amount The amount locked.
+    event Locked(address origin, bytes32 recipient, address token, uint128 amount);
+
+    /// @dev Emitted once the funds are unlocked.
+    /// @param origin The substrate address which initiated the unlock.
+    /// @param recipient The ethereyn address that will receive the funds.
+    /// @param token The token unlocked.
+    /// @param amount The amount unlocked.
+    event Unlocked(bytes32 origin, address recipient, address token, uint128 amount);
+
+    /// @dev Emitted after enqueueing a a create token message to substrate.
+    /// @param token The address of the token created.
+    // TODO: Remove name, symbol and decimals.
+    event Created(address token, string name, string symbol, uint8 decimals);
+
+    /// @dev Zero amount.
+    error ZeroAmount();
+
+    /// @dev Zero addressed token.
+    error ZeroAddressToken();
+
+    /// @dev Zero addressed recipient.
+    error ZeroAddressRecipient();
+
+    /// @dev the origin provided is unauthorized.
+    error UnauthorizedOrigin();
+
+    /// @dev the message action is unsupported.
+    error UnsupportedMessageAction();
+
+    /// @dev the origin that
+    bytes32 public immutable allowedOrigin;
+
+    /// @dev The vault where ERC20 tokens are locked.
+    ERC20Vault public immutable vault;
+
+    /// @dev The channel used to enqueue messages for lock and create functions.
+    OutboundChannel public immutable outboundChannel;
+
+    /// @dev Initializes the NativeTokens contract with a vault and channels.
+    /// @param _vault The vault to use to `lock`/`unlock` tokens.
+    /// @param _outboundChannel The channel used to queue lock and create messages.
+    /// @param _allowedOrigin The origin allowed to call handle.
+    constructor(ERC20Vault _vault, OutboundChannel _outboundChannel, bytes32 _allowedOrigin) {
```

Comment:
We need to resurrect the `ChannelRegistry` contract. Is necessary to allow the outbound channel to be upgraded. Can you pull it from our git history?

Reply:
I will create a new PR for this.